lib/msal-browser/src/broker/EmbeddedClientApplication.ts

```diff
@@ -4,72 +4,107 @@
+import { BrowserStorage } from "../cache/BrowserStorage";
 import { BrowserAuthError } from "../error/BrowserAuthError";
 
 const DEFAULT_MESSAGE_TIMEOUT = 2000;
 /**
  * Embedded application in a broker scenario.
  */
-export class BrokerClient {
+export class EmbeddedClientApplication {
     private brokerOptions: BrokerOptions;
     private logger: Logger;
     private clientId: string;
     private version: string;
     private brokerOrigin: string;
+    private browserStorage: BrowserStorage;
 
+    private get trustedBrokersProvided(): boolean {
+        return this.brokerOptions.trustedBrokerDomains && this.brokerOptions.trustedBrokerDomains.length >= 1;
+    };
     public brokeringEnabled: boolean;
 
-    constructor(brokerOptions: BrokerOptions, logger: Logger, clientId: string, version: string) {
+    constructor(brokerOptions: BrokerOptions, logger: Logger, clientId: string, version: string, browserStorage: BrowserStorage) {
         this.brokerOptions = brokerOptions;
         this.logger = logger;
         this.clientId = clientId;
         this.version = version;
-
-        if (!this.brokerOptions.trustedBrokerDomains || this.brokerOptions.trustedBrokerDomains.length < 1) {
-            this.logger.info("Application was identified as an embedded app in a broker scenario, but no trusted broker domains were provided.");
-            this.brokeringEnabled = false;
-        } else {
-            this.brokeringEnabled = true;
-        }
+        this.browserStorage = browserStorage;
+        this.brokeringEnabled = false;
     }
 
     async initiateHandshake(): Promise<void> {
-        if (!this.brokeringEnabled) {
-            this.logger.verbose("Brokering is not enabled, handshake was not performed.");
-            return;
+        if (!this.trustedBrokersProvided) {
+            throw BrowserAuthError.createNoTrustedBrokersProvidedError();
         }
 
         try {
             const response = await this.sendHandshakeRequest();
             this.brokerOrigin = response.brokerOrigin;
+            this.brokeringEnabled = true;
         } catch (e) {
             this.logger.error(e);
             this.brokeringEnabled = false;
             throw e;
         }
     }
 
+    private async preflightBrokerRequest(): Promise<void> {
+        if (!this.brokeringEnabled) {
+            this.logger.info("Attempting handshake...");
+            try {
+                await this.initiateHandshake();
+            } catch (e) {
+                this.logger.error("Handshake rejected");
+                throw BrowserAuthError.createBrokeringDisabledError(e);
+            }
+        }
+    }
+
     async sendPopupRequest(request: PopupRequest): Promise<AuthenticationResult> {
-        return this.sendRequest<AuthenticationResult>(request, InteractionType.POPUP, 6000);
+        await this.preflightBrokerRequest();
+
+        const brokerAuthResultMessage = await this.sendRequest(request, InteractionType.POPUP, 60000);
```

Comment:
Please define this number somewhere and reference it. even if just at the top of the file
```js
const POPUP_TIMEOUT_LIMIT = 6000
```

Reply:
Should this be configurable?

Reply:
done.